File: scripts/sync_to_sheets.mjs
#!/usr/bin/env node
// Run: node scripts/sync_to_sheets.mjs
// Syncs all reservation_guests from last 10 days to Google Sheets (with redirect follow)
import { createRequire } from 'module';
import { fileURLToPath } from 'url';
import path from 'path';
import https from 'https';
import http from 'http';

const require = createRequire(import.meta.url);
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const Database = require('better-sqlite3');

const DB_PATH = path.join(__dirname, '../data/alisio.db');
const GOOGLE_SCRIPT_URL = process.env.GOOGLE_GUESTS_SCRIPT_URL || '';
const DAYS_BACK = parseInt(process.env.DAYS_BACK || '10', 10);

if (!GOOGLE_SCRIPT_URL) {
  console.error('ERROR: GOOGLE_GUESTS_SCRIPT_URL not set');
  process.exit(1);
}

const db = new Database(DB_PATH);

const rows = db.prepare(`
  SELECT rg.*, r.check_in, r.check_out
  FROM reservation_guests rg
  JOIN reservations r ON rg.reservation_id = r.id
  WHERE rg.created_at >= datetime('now', '-' || ? || ' days')
  ORDER BY rg.created_at ASC
`).all(DAYS_BACK);

console.log(`[sheets] Found ${rows.length} guests to sync to Google Sheets`);

function postWithRedirects(url, payload, maxRedirects = 5) {
  return new Promise((resolve) => {
    const body = JSON.stringify(payload);
    const parsed = new URL(url);
    const lib = parsed.protocol === 'https:' ? https : http;

    const options = {
      hostname: parsed.hostname,
      path: parsed.pathname + parsed.search,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
      },
    };

    const req = lib.request(options, (res) => {
      if ([301, 302, 303, 307, 308].includes(res.statusCode) && res.headers.location && maxRedirects > 0) {
        // Follow redirect — Google Apps Script always redirects
        const redirectUrl = res.headers.location.startsWith('http')
          ? res.headers.location
          : `${parsed.protocol}//${parsed.hostname}${res.headers.location}`;

        // After redirect, use GET (as per 302 semantics) or POST again
        // Google Apps Script needs GET after redirect
        const rParsed = new URL(redirectUrl);
        const rLib = rParsed.protocol === 'https:' ? https : http;
        const rReq = rLib.request({
          hostname: rParsed.hostname,
          path: rParsed.pathname + rParsed.search,
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(body),
          },
        }, (rRes) => {
          let data = '';
          rRes.on('data', c => data += c);
          rRes.on('end', () => resolve({ status: rRes.statusCode, body: data.slice(0, 200) }));
        });
        rReq.on('error', e => resolve({ status: 0, body: e.message }));
        rReq.write(body);
        rReq.end();
        res.resume();
        return;
      }
      let data = '';
      res.on('data', c => data += c);
      res.on('end', () => resolve({ status: res.statusCode, body: data.slice(0, 200) }));
    });
    req.on('error', e => resolve({ status: 0, body: e.message }));
    req.write(body);
    req.end();
  });
}

let sent = 0;
let errors = 0;

for (const row of rows) {
  const nights = row.check_in && row.check_out
    ? Math.max(0, (new Date(row.check_out + 'T00:00:00Z') - new Date(row.check_in + 'T00:00:00Z')) / 86400000)
    : 0;

  const payload = {
    action: 'guest',
    full_name: `${row.last_name} ${row.first_name}`.trim(),
    surname: row.last_name || '',
    first_name: row.first_name || '',
    birth_date: row.date_of_birth || '',
    doc_type: row.document_type || '',
    doc_number: row.document_number || '',
    country_code: '',
    nationality: row.nationality || '',
    address: row.address || '',
    visa_number: '',
    check_in: row.check_in || '',
    check_out: row.check_out || '',
    nights,
    is_foreigner: 'Tak',
    tax_amount: 0,
    exempt_reason: '',
    purpose: '',
    note: '[Migrated from guest portal web registration]',
  };

  const result = await postWithRedirects(GOOGLE_SCRIPT_URL, payload);
  if (result.status === 200 && !result.body.toLowerCase().includes('error')) {
    sent++;
    console.log(`[sheets] ✅ ${payload.full_name} | ${payload.check_in}`);
  } else {
    errors++;
    console.error(`[sheets] ❌ ${payload.full_name} → status=${result.status} | ${result.body.slice(0, 80)}`);
  }
  await new Promise(r => setTimeout(r, 700));
}

console.log(`\n[sheets] Done: ${sent} sent, ${errors} errors`);

File: scripts/sync_to_sheets.py
#!/usr/bin/env python3
"""
Sync all reservation_guests from last 10 days to Google Sheets.
Uses the same aiohttp/requests approach as the Telegram bot.
Run: python3 scripts/sync_to_sheets.py
"""
import json
import os
import sys
import time
import sqlite3
import urllib.request
import urllib.error
from datetime import datetime, timedelta

DB_PATH = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'data', 'alisio.db')
GOOGLE_URL = os.environ.get('GOOGLE_GUESTS_SCRIPT_URL', '')
DAYS_BACK = int(os.environ.get('DAYS_BACK', '10'))

if not GOOGLE_URL:
    print('ERROR: GOOGLE_GUESTS_SCRIPT_URL not set')
    sys.exit(1)

conn = sqlite3.connect(DB_PATH)
conn.row_factory = sqlite3.Row
cur = conn.cursor()

since = (datetime.now() - timedelta(days=DAYS_BACK)).strftime('%Y-%m-%d %H:%M:%S')
cur.execute('''
    SELECT rg.id, rg.first_name, rg.last_name, rg.document_number, rg.document_type,
           rg.nationality, rg.date_of_birth, rg.address, rg.created_at,
           r.check_in, r.check_out
    FROM reservation_guests rg
    JOIN reservations r ON rg.reservation_id = r.id
    WHERE rg.created_at >= ?
    ORDER BY rg.created_at ASC
''', (since,))

rows = cur.fetchall()
print(f'[sheets] Found {len(rows)} guests to sync (since {since})')

sent = 0
errors = 0

for row in rows:
    try:
        ci = datetime.strptime(row['check_in'], '%Y-%m-%d') if row['check_in'] else None
        co = datetime.strptime(row['check_out'], '%Y-%m-%d') if row['check_out'] else None
        nights = max(0, (co - ci).days) if ci and co else 0
    except Exception:
        nights = 0

    payload = {
        'action': 'guest',
        'full_name': f"{row['last_name']} {row['first_name']}".strip(),
        'surname': row['last_name'] or '',
        'first_name': row['first_name'] or '',
        'birth_date': row['date_of_birth'] or '',
        'doc_type': row['document_type'] or '',
        'doc_number': row['document_number'] or '',
        'country_code': '',
        'nationality': row['nationality'] or '',
        'address': row['address'] or '',
        'visa_number': '',
        'check_in': row['check_in'] or '',
        'check_out': row['check_out'] or '',
        'nights': nights,
        'is_foreigner': 'Tak',
        'tax_amount': 0,
        'exempt_reason': '',
        'purpose': '',
        'note': '[Migrated from guest portal web registration]',
    }

    body = json.dumps(payload).encode('utf-8')
    req = urllib.request.Request(
        GOOGLE_URL,
        data=body,
        headers={'Content-Type': 'application/json'},
        method='POST',
    )

    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            resp_body = resp.read().decode('utf-8', errors='ignore')[:200]
            if resp.status == 200 and 'error' not in resp_body.lower():
                sent += 1
                print(f"[sheets] ✅ {payload['full_name']} | {payload['check_in']} → {payload['check_out']}")
            else:
                errors += 1
                print(f"[sheets] ❌ {payload['full_name']} status={resp.status} body={resp_body[:80]}")
    except urllib.error.HTTPError as e:
        errors += 1
        print(f"[sheets] ❌ HTTP {e.code} for {payload['full_name']}: {e.reason}")
    except Exception as e:
        errors += 1
        print(f"[sheets] ❌ ERROR for {payload['full_name']}: {e}")

    time.sleep(0.7)

print(f'\n[sheets] Done: {sent} sent, {errors} errors')
conn.close()
